service-base/src/main/java/org/eclipse/hono/service/monitoring/ConnectionEventProducer.java

```diff
@@ -0,0 +1,49 @@
+/**
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 1.0 which is available at
+ * https://www.eclipse.org/legal/epl-v10.html
+ *
+ * SPDX-License-Identifier: EPL-1.0
+ */
+package org.eclipse.hono.service.monitoring;
+
+import org.eclipse.hono.service.auth.device.Device;
+
+/**
+ * Produces connection events.
+ * <p>
+ * The interface is intended to be implemented by setups which are interested in receiving technical connection events.
+ * Which might not make sense for all protocols adapters. But mostly for connection oriented ones like MQTT or AMQP 1.0.
+ * <p>
+ * The protocol adapters may call {@link #connected(String, Device)} and {@link #disconnected(String, Device)} as they
+ * see fit. The whole process is a "best effort" process and intended to be used for monitoring/debugging.
+ * <p>
+ * When a protocol adapter calls into this producer it must provide some kind of connection ID, which might have a
+ * different meaning for different protocol adapters. This should be as identifiable as possible but there is no
+ * requirement for this to be unique. However for each connection the protocol adapter must call
+ * {@link #disconnected(String, Device)} with the same information as it called {@link #connected(String, Device)}.
+ */
+public interface ConnectionEventProducer {
```

Comment:
I don't see why we should introduce this new concept when we already have an event propagation mechanism. FMPOV we should simply send an event using the Event API downstream. It would be sufficient to simply define the event(s) and use what we already have.

Reply:
I agree to @sophokles73 - sending a special purpose event downstream is what we discussed in the PR #562.
And it is what is now implemented in the PR #576 regarding the explicity sent *device notification*s - just using the Event API, but use it with a well defined event type, supported by a specific class.

I propose we do something similar here - define a class that is JSON mappable by jackson and send the payload via the Event API. 

After having implemented this for the devices, I would next go back to the specification and include these *protocol adpater connection events* as well. So then we have a stable specification that you could follow to integrate it for the MQTT protocol adapter.
Does that make sense?

Reply:
Well the idea was to provide some kind of pluggability here. Instead of directly hooking in the Event API allow different approaches here.

As I tried to explain in #565, capture the locations triggering the event, and then providing a way to implement the handling. The default handling could be what @sysexcontrol described, using the Event API to send a specialized event downstream.

Adding an interface and allowing an alternative implementation would allow consumers of Hono to route such events to a different address (e.g. Alerts) or disable them completely. If we directly invoke the Event API that won't be possible.

Reply:
Regarding *pluggability* :+1: 

But since we implemented our microservices with the Spring framework, I would prefer to use what Spring offers for pluggability, i.e. declaring a `@Bean` and autowire it to the necessary classes.
This is as flexible as it can be: we autowire a default `@Bean` (using the Event API), may provide some more (like your logging only implementation) and can easily grow in the future.
WDYT?

Reply:
I completely agree. But I think that is already present in the PR. I extended the `Config` class of the protocol adapter and used the `@Autowired` annotation.